rust/src/acl/mod.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of Matrix server ACL rules.
+
+use crate::push::utils::{glob_to_regex, GlobMatchType};
+use anyhow::Error;
+use pyo3::prelude::*;
+use regex::Regex;
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+
+/// Called when registering modules with python.
+pub fn register_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    let child_module = PyModule::new(py, "acl")?;
+    child_module.add_class::<ServerAclEvaluator>()?;
+
+    m.add_submodule(child_module)?;
+
+    // We need to manually add the module to sys.modules to make `from
+    // synapse.synapse_rust import acl` work.
+    py.import("sys")?
+        .getattr("modules")?
+        .set_item("synapse.synapse_rust.acl", child_module)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+#[pyclass(frozen)]
+pub struct ServerAclEvaluator {
+    allow_ip_literals: bool,
+    allow: Vec<Regex>,
+    deny: Vec<Regex>,
+}
+
+#[pymethods]
+impl ServerAclEvaluator {
+    #[new]
+    pub fn py_new(
+        allow_ip_literals: bool,
+        allow: Vec<String>,
+        deny: Vec<String>,
+    ) -> Result<Self, Error> {
+        let allow = allow
+            .iter()
+            .map(|s| glob_to_regex(s, GlobMatchType::Whole).unwrap())
```

Comment:
Do we normally unwrap glob_to_regex's output? I guess a panic-on-unwrap gets turned into a Python exception of some kind?

(I was just a bit surprised that would panic rather than use `?` to return the error.)

Reply:
I imagine this is because getting errors out of iterators is non obvious. The trick to it is to do:

```rust
let allow = allow
            .iter()
            .map(|s| glob_to_regex(s, GlobMatchType::Whole))
            .collect::<Result<_, _>>()?;
```

As `collect()` will correctly convert an iterator of results into a `Result<Collection, _>`, using the first error it encounters. (This is because of [this `FromIterator` implementation](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E)).

An existing example of us doing this is:

https://github.com/matrix-org/synapse/blob/2763c49eca483dbb848b70b951891afd57016f17/rust/build.rs#L17-L19

Reply:
Thanks @erikjohnston! That looks much better.